lib/document.js

```diff
@@ -1394,7 +1394,7 @@ Document.prototype.$set = function $set(path, val, type, options) {
 
     let didPopulate = false;
     if (refMatches && val instanceof Document && (!val.$__.wasPopulated || utils.deepEqual(val.$__.wasPopulated.value, val._id))) {
```

Comment:
Not completely sure, but did you miss the `_doc` in the `deepEqual` check?

```diff
-    if (refMatches && val instanceof Document && (!val.$__.wasPopulated || utils.deepEqual(val.$__.wasPopulated.value, val._id))) {
+    if (refMatches && val instanceof Document && (!val.$__.wasPopulated || utils.deepEqual(val.$__.wasPopulated.value, val._doc._id))) {
```

Reply:
Otherwise we run into this situation

```js
const owner = new Owner({ name: 'Alice' })
const pet1 = new Pet({ name: 'Kitty1', owner: owner })
const pet2 = new Pet({ name: 'Kitty2', owner: owner })

console.log(typeof pet1.owner) // 'object'
console.log(typeof pet2.owner) // 'string'
```

Reply:
Fixed in 1ad7531, thanks for catching this